z3c/dependencychecker/db.py

```diff
@@ -43,3 +48,186 @@ def _filter_duplicates(self, imports):
+        for extra in self._extras_requirements:
+            all_but_test_requirements.update(self._extras_requirements[extra])
+        test_requirements = self._get_test_extra()
+        if test_requirements:
+            for dotted_name in test_requirements:
+                all_but_test_requirements.remove(dotted_name)
+        filters = (
+            self._filter_known_packages,
+            self._filter_python_standard_library,
+            self._filter_used_imports,
+        )
+        unneeded = self._process_pipeline(all_but_test_requirements, filters)
+        unique_imports = self._get_unique_imports(imports_list=unneeded)
+        return unique_imports
+
+    def requirements_that_should_be_test_requirements(self):
+        filters = (
+            self._filter_testing_imports,
+            self._filter_python_standard_library,
+            self._filter_own_package,
+        )
+        non_testing_imports = self._process_pipeline(
+            self.imports_used,
+            filters,
+        )
+        requirements_not_used = [
+            x
+            for x in self._requirements
+            if self._filter_requirements_lists(non_testing_imports, x)
+        ]
+        filters = (
+            self._filter_only_testing_imports,
+            self._filter_python_standard_library,
+            self._filter_own_package,
+        )
+        testing_imports = self._process_pipeline(
+            self.imports_used,
+            filters,
+        )
+        should_be_test_requirements = [
+            x
+            for x in requirements_not_used
+            if not self._filter_requirements_lists(testing_imports, x)
+        ]
+        unique = self._get_unique_imports(
+            imports_list=should_be_test_requirements,
+        )
+        return unique
+
+    def get_unneeded_test_requirements(self):
+        test_requirements = self._get_test_extra()
+        if not test_requirements:
+            return []
+
+        filters = (
+            self._filter_known_packages,
+            self._filter_python_standard_library,
+            self._filter_used_imports,
+        )
+        unneeded = self._process_pipeline(test_requirements, filters)
+        unique_imports = self._get_unique_imports(imports_list=unneeded)
+        return unique_imports
+
+    def _process_pipeline(self, objects, filters):
+        final_list = []
+        for single_object in objects:
+            for filter_func in filters:
+                if not filter_func(single_object):
+                    break
+            else:
+                final_list.append(single_object)
+
+        return final_list
+
+    def _get_unique_imports(self, imports_list=None):
+        if imports_list is None:
+            imports_list = self.imports_used
+        unique_dotted_names = list(set(imports_list))
+        sorted_unique_dotted_names = sorted(unique_dotted_names)
+        return sorted_unique_dotted_names
+
+    def _filter_used_imports(self, dotted_name):
```

Comment:
What do you think about renaming those methods from `_filter_something()` to `_filter_out_something()`?

The methods return True or False. Having a "negative" method name makes it harder to interpret "False". "do_not_include=False", what does that mean?

Hm. I also see `_filter_testing_imports()` that returns True/False when the dotted name is a test or not. But `_filter_known_packages()` returns a list of requirements.

=> **Something** has to be changed regarding the naming, I think?

Reply:
``_filter_requirements_lists`` is a bit involved, but all ``_filter_XXX`` methods do return either ``True`` or ``False`` (regarding ``_filter_testing_imports`` and ``_filter_known_packages``).

Regarding renaming to ``filter_out``, sure no problem.